feat: add method to retrieve bindingObserver subscriptions

nicholasrice opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting. Someone may have pushed the same thing before!

Provide a summary of your changes in the title field above.
-->

# Pull Request

## 📖 Description
This PR adds a method to `BindingObserver` that allows inspection of the internal `SubscriptionRecord`'s used being used by a `BindingObserver`. This allows us to actually inspect which objects & values are being observed by a `BindingObserver`.

<!---
Provide some background and a description of your work.
What problem does this change solve?
Is this a breaking change, chore, fix, feature, etc?
-->

### 🎫 Issues

<!---
* List and link relevant issues here.
-->

## 👩‍💻 Reviewer Notes

<!---
Provide some notes for reviewers to help them provide targeted feedback and testing.

Do you recommend a smoke test for this PR? What steps should be followed?
Are there particular areas of the code the reviewer should focus on?
-->

## 📑 Test Plan

<!---
Please provide a summary of the tests affected by this work and any unique strategies employed in testing the features/fixes.
-->

## ✅ Checklist

### General

<!--- Review the list and put an x in the boxes that apply. -->

- [ ] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

<!--- Review the list and put an x in the boxes that apply. -->
<!--- Remove this section if not applicable. -->

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## ⏭ Next Steps

<!---
If there is relevant follow-up work to this PR, please list any existing issues or provide brief descriptions of what you would like to do next.
-->

nicholasrice commented on packages/web-components/fast-element/src/observation/observable.ts:
```diff
@@ -377,6 +377,12 @@ export interface BindingObserver<TSource = any, TReturn = any, TParent = any>
+     * Gets {@link SubscriptionRecord|SubscriptionRecords} that the {@link BindingObserver}
+     * is observing.
+     */
+    subscriptionRecords(): IterableIterator<SubscriptionRecord>;
```
@EisenbergEffect I wasn't exactly sure what to name this so if you have something better in mind let me know. I also thought about simply `BindingObserver.records()`

EisenbergEffect replied:
Let's go with `records` if that's ok with you.

EisenbergEffect replied:
Also, I think we may need to export `SubscriptionRecord` or some version of it. Do we want to export that directly or have a version that we export with only the `propertySource` and `propertyName` properties? Should we include `notifier` as well?

nicholasrice replied:
The the `notifier` can get obtained from `Observable.getNotifier(propertySource)` right? If so I think we can probably just expose `propertySource` and `propertyName`

nicholasrice replied:
@EisenbergEffect changed method name to `records()` and exported an `ObservationRecord` interface

EisenbergEffect reviewed (changes requested):
Looks good. Just had one point I wanted us to ponder before merging.

EisenbergEffect reviewed (approved):
Looks good. If you can, add some docs on the new interface.

nicholasrice commented:
> Looks good. If you can, add some docs on the new interface.

Added some interface docs and updated observable docs for doc site. Lemme know if that all sounds right or not.

nicholasrice commented on packages/web-components/fast-element/docs/guide/observables-and-state.md:
```diff
@@ -191,4 +191,20 @@ const handler = {
+const value = bindingObserver.observe({}, defaultExecutionContext);
+
+for (const record of bindingObserver.records()) {
+  // Do something with the bindings observable dependencies
```
```suggestion
  // Do something with the binding's observable dependencies
```

EisenbergEffect commented:
Thanks @nicholasrice !
